a10networks/a10-octavia
# [Feature][STACK-2017][vThunder support] - a10-octavia vThunder flow support

## Description
Before a10-octavia support vThunder flows, which a10-octavia can automatically create vThunder instance when loadbalancer is created in a openstack project. But the flow didn’t be maintained for a while. For customer requirements a10-octavia v1.2 will support vThunder as amphora with most features. 

Supported features:
- basic functionality
- VRRP-A floating ip support
- Flavor support
- SNAT
- write-memory enhancement

Design document: 
https://teams.microsoft.com/l/file/96CCF61A-DE50-4BEB-9B14-6A4D215270C3?tenantId=91d27ab9-8c5e-41d4-82e8-3d1bf81fcb2f&fileType=docx&objectUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack%2FShared%20Documents%2FDevelopment%2FResearch%20%26%20Design%2Fa10-octavia%20v1.2%2FvThunder%20support%2F%5BDD%5D%20vThunder%20support.docx&baseUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack&serviceName=teams&threadId=19:0f1acbb173d74758b05ee8bacb000d68@thread.tacv2&groupId=37bbf3ad-c05a-4e67-b0cc-12bcd1479beb

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-2017
https://a10networks.atlassian.net/browse/STACK-2016

## Technical Approach
- Move features in hardware thunder flow to vThunder flow

## Config Changes
Please reference following document for vThunder configuration and setup:
https://teams.microsoft.com/l/file/CED51F5A-0A31-4B09-8789-934939826F58?tenantId=91d27ab9-8c5e-41d4-82e8-3d1bf81fcb2f&fileType=docx&objectUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack%2FShared%20Documents%2FDevelopment%2FResearch%20%26%20Design%2Fa10-octavia%20v1.2%2FvThunder%20support%2Fvthunder%20setup.docx&baseUrl=https%3A%2F%2Fa10networks.sharepoint.com%2Fsites%2FOpenstack&serviceName=teams&threadId=19:0f1acbb173d74758b05ee8bacb000d68@thread.tacv2&groupId=37bbf3ad-c05a-4e67-b0cc-12bcd1479beb


## Test Cases
Tested Features:
- basic functionality (tested by automation)
- VRRP-A floating ip support (tested by Adeeb)
- Flavor support (tested by Neha)
- SNAT (tested by Neha)
- write-memory enhancement (tested by Adeeb)

## Manual Testing
Please check PRs for detail test items and results:
- basic operations: https://github.com/a10networks/a10-octavia/pull/322
- VCS and Active-Standby: https://github.com/a10networks/a10-octavia/pull/327
- VRID and VRRP-A Floating-IP support: https://github.com/a10networks/a10-octavia/pull/335
- Flavor Support: https://github.com/a10networks/a10-octavia/pull/323
- SNAT: https://github.com/a10networks/a10-octavia/pull/321
- Periodically Write Memory: https://github.com/a10networks/a10-octavia/pull/325
- Periodically Write Memory for Active-Standby topology: https://github.com/a10networks/a10-octavia/pull/330
